### ISSUE TITLE
Debugger pauses at end of Dart test run

### ISSUE THREAD

**Issue by DanTup** [flutter/flutter-intellij#5043]
I'm not sure if this is the right place for this (it's Dart, not Flutter), but in Android Studio using a recent nightly build, debugging a test will pause at the end of the rest run. This is the same issue described at https://github.com/Dart-Code/Dart-Code/issues/2949.

I've added a workaround in Dart-Code for now (when the `done` event comes through, disable break-on-exceptions, remove breakpoints, and resume all threads - we already had this code that we run when the user clicks the Stop button on the debug toolbar), though it's possible Pub could get a better fix.

<img width="810" alt="Screenshot 2020-11-12 at 15 22 43" src="https://user-images.githubusercontent.com/1078012/98959569-50726e80-24fb-11eb-859a-1aaf7825c3d8.png">

@devoncarew

**Comment by devoncarew** [comment]
@grouma @natebosch - are you aware of any recent changes in the test runner wrt it now terminating with an `ExitWithException`? Or any changes in terms of the workflow around that?

I'm wondering if this is a change in package:test, in whether this exception is now considered uncaught, in some VM behavior? It would be great for all IDEs to not have to work around this - to not break on what's likely just package:test infrastructure and not user test code.

**Comment by DanTup** [comment]
There's a little info from @sigurdm at https://github.com/Dart-Code/Dart-Code/issues/2949#issuecomment-724791777.

Migrating to `dart test` may be the best end-goal, though right now that doesn't seem to work right with the options used in Dart-Code (I've an open issue at https://github.com/Dart-Code/Dart-Code/issues/2950) so I'm not sure what the best short-term route is.

**Comment by devoncarew** [comment]
cc @alexander-doroshko for FYI

@DanTup - from the conversation on that Dart-Code issue:
- is there a good workaround for the behavior change?
- should we revert the change to use exceptions for control flow?
- is there a path forward with `dart test` (or do we need changes there)?

**Comment by DanTup** [comment]
> is there a good workaround for the behavior change?

My workaround is to call code that I already had for when the user clicks Stop in the debugger (we want to shut down, which means we do not want to pause the debugger, so I set don't-break-on-exceptions, remove all breakpoints, and resume all isolates). It works, however there's a very fringe case that if the user had some code that ran after the test run ended that throws, we also would not break on that anymore.

> should we revert the change to use exceptions for control flow?

That's probably a question for @sigurdm. I'm not familiar with the motivation for the change or the consequences of reverting.

> is there a path forward with `dart test` (or do we need changes there)?

Currently there seem to be blockers (see https://github.com/Dart-Code/Dart-Code/issues/2950) with using `--pause_isolates_on_start=true` and/or `--write-service-info`. If the intention is that IDEs will move over, they will need to be addressed - or alternatives provided - though I'm not sure where that lies (`pub` or VM). If you (or @sigurdm know where that likely is), I can file an issue with some more details.

**Comment by natebosch** [comment]
> are you aware of any recent changes in the test runner wrt it now terminating with an `ExitWithException`? Or any changes in terms of the workflow around that?

That isn't a `package:test` thing, it's a `pub` thing. @sigurdm 

https://github.com/dart-lang/pub/commit/63b56ea42fc58dd7eafa83c5b3c0b05ffdef0712

**Comment by DanTup** [comment]
@devoncarew turns out there's an easy fix for this (thanks @sigurdm/@jonasfj!) by using `dart run test:test` instead of `dart [pub-snapshot] run test`.

The Dart-Code change is here (I've also reverted the previous workaround) and is gated on v2.12 of the SDK:

https://github.com/Dart-Code/Dart-Code/pull/2956/commits/c7f81d6324a819946609ffca0ef1956a21c5a771

**Comment by devoncarew** [comment]
Hi, catching up on this:

- I don't think we want to have to update existing clients for the new pause-at-test-completion behavior. This is a regression from the POV of users in running tests. I'd like to revert the change that uses exceptions for control flow here; I'll open a bug on the dart sdk about that
- I think that in the 2.12.0 timeframe, instead of working around an issue w/ passed VM flags via using `dart run test:test`, we should address any args parsing issues so that both users and IDEs can use the new `dart test` command; I'll open abother issue about that

**Comment by devoncarew** [comment]
I filed https://github.com/dart-lang/pub/issues/2753 for changing back to pub's previous behavior (for existing clients), and https://github.com/dart-lang/sdk/issues/44200 for improving `dart test` enough so that IDEs can start using it for running and debugging tests in the 2.12.0 timeframe.

**Comment by DanTup** [comment]
Sounds reasonable to me. I've added some notes to https://github.com/dart-lang/sdk/issues/44200 with the command that works and the one that doesn't.

I've already reverted the original workaround in Dart-Code, so if/when `dart test` supports those flags, I'll update the change linked above to use `dart test` in place of `dart run test:test`.

(if both of those can/will be done, I think there's nothing to do here then)
